plugins/pipelines/badge.js

```diff
@@ -54,64 +55,71 @@ module.exports = config => ({
                     }
-
-                    return pipeline.getEvents({ sort: 'ascending' }).then(allEvents => {
-                        const getLastEffectiveEvent = events => {
-                            const lastEvent = events.pop();
-
-                            if (!lastEvent) {
-                                return h.response(getPipelineBadge(badgeConfig)).header('Content-Type', contentType);
-                            }
-
-                            return lastEvent.getBuilds().then(builds => {
-                                if (!builds || builds.length < 1) {
-                                    return getLastEffectiveEvent(events);
-                                }
-
-                                const buildsStatus = builds.reverse().map(build => build.status.toLowerCase());
-
-                                let workflowLength = 0;
-
-                                if (lastEvent.workflowGraph) {
-                                    const nextJobs = dfs(lastEvent.workflowGraph, lastEvent.startFrom, lastEvent.prNum);
-
-                                    workflowLength = nextJobs.size;
-                                }
-
-                                for (let i = builds.length; i < workflowLength; i += 1) {
-                                    buildsStatus[i] = 'unknown';
-                                }
-
-                                return h
-                                    .response(
-                                        getPipelineBadge(
-                                            Object.assign(badgeConfig, {
-                                                buildsStatus,
-                                                label: pipeline.name
-                                            })
-                                        )
-                                    )
-                                    .header('Content-Type', contentType);
-                            });
-                        };
-
-                        return getLastEffectiveEvent(allEvents);
-                    });
-                })
-                .catch(() => h.response(getPipelineBadge(badgeConfig)));
+                });
+
+                if (!latestEvents || Object.keys(latestEvents).length === 0) {
+                    return h.response(getPipelineBadge(badgeConfig)).header('Content-Type', contentType);
+                }
+
+                // Only care about latest
+                const lastEvent = latestEvents[0];
+                const builds = await lastEvent.getBuilds({ readOnly: true });
+
+                if (!builds || builds.length < 1) {
+                    return h.response(getPipelineBadge(badgeConfig)).header('Content-Type', contentType);
+                }
+
+                const buildsStatus = builds.reverse().map(build => build.status.toLowerCase());
+                // Get downstream jobs
+                const nextJobs = dfs(lastEvent.workflowGraph, lastEvent.startFrom, lastEvent.prNum);
```

Comment:
Is it intentional that `if (lastEvent.workflowGraph) {}` check is removed?

Reply:
yeah I think workflowgraph will always exist
but ended up removing that entire section